client/state.go

```diff
@@ -0,0 +1,307 @@
+	}
+)
+
+// State is a collection of StateItems with a store name
+type State struct {
+	StoreName string
+	States    []*StateItem
+}
+
+// StateItem represents the state to be persisted
+type StateItem struct {
+	Key      string
+	Value    []byte
+	Etag     string
+	Metadata map[string]string
+	Options  *StateOptions
+}
+
+// StateOptions represents the state store persistence policy
+type StateOptions struct {
+	Concurrency StateConcurrency
+	Consistency StateConsistency
+	RetryPolicy *StateRetryPolicy
+}
+
+// StateRetryPolicy represents the state store invocation retry policy
+type StateRetryPolicy struct {
+	Threshold int32
+	Pattern   RetryPattern
+	Interval  time.Duration
+}
+
+// *** Converters
+
+func toProtoSaveStateRequest(s *State) (req *pb.SaveStateRequest) {
+	r := &pb.SaveStateRequest{
+		StoreName: s.StoreName,
+		States:    make([]*v1.StateItem, 0),
+	}
+
+	for _, si := range s.States {
+		item := toProtoSaveStateItem(si)
+		r.States = append(r.States, item)
+	}
+	return r
+}
+
+func toProtoSaveStateItem(si *StateItem) (item *v1.StateItem) {
+	return &v1.StateItem{
+		Etag:     si.Etag,
+		Key:      si.Key,
+		Metadata: si.Metadata,
+		Value:    si.Value,
+		Options:  toProtoStateOptions(si.Options),
+	}
+}
+
+func toProtoStateOptions(so *StateOptions) (opts *v1.StateOptions) {
+	if so == nil {
+		return stateOptionDefault
+	}
+	return &v1.StateOptions{
+		Concurrency: (v1.StateOptions_StateConcurrency(so.Concurrency)),
+		Consistency: (v1.StateOptions_StateConsistency(so.Consistency)),
+		RetryPolicy: &v1.StateRetryPolicy{
+			Interval:  toProtoDuration(so.RetryPolicy.Interval),
+			Pattern:   (v1.StateRetryPolicy_RetryPattern(so.RetryPolicy.Pattern)),
+			Threshold: so.RetryPolicy.Threshold,
+		},
+	}
+}
+
+func toProtoDuration(d time.Duration) *duration.Duration {
+	nanos := d.Nanoseconds()
+	secs := nanos / 1e9
+	nanos -= secs * 1e9
+	return &duration.Duration{
+		Seconds: int64(secs),
+		Nanos:   int32(nanos),
+	}
+}
+
+// *** Save State ***
+
+// SaveState saves the fully loaded save state request
+func (c *Client) SaveState(ctx context.Context, s *State) error {
+	if s == nil || s.StoreName == "" || s.States == nil || len(s.States) < 1 {
+		return errors.New("nil or invalid state")
+	}
+	req := toProtoSaveStateRequest(s)
+	_, err := c.protoClient.SaveState(authContext(ctx), req)
+	if err != nil {
+		return errors.Wrap(err, "error saving state")
+	}
+	return nil
+}
+
+// SaveStateItem saves a single state item
+func (c *Client) SaveStateItem(ctx context.Context, store string, item *StateItem) error {
+	if store == "" {
+		return errors.New("nil store")
+	}
+	if item == nil {
+		return errors.New("nil item")
+	}
+
+	req := &State{
+		StoreName: store,
+		States:    []*StateItem{item},
+	}
+
+	return c.SaveState(ctx, req)
+}
+
+// SaveStateWithData saves the data into store using default state options
+func (c *Client) SaveStateWithData(ctx context.Context, store, key string, data []byte) error {
```

Comment:
Can it take an ETag as well, else for Saving state with ETag user will always have to create a StateItem and then call SaveStateItem.
Can we look into renaming and reorganizing these method names. For Delete there is DeleteStateWithOptions & DeleteState. For SaveState, I would expect similar methods.

Reply:
We can certainly think through the method naming. IMO the diff in between Delete and Save is that Delete does have an option to delete using state options in the proto. In case of Save, there is Save multiple items, in which case each item has its own state option. I added Save one item to not force user to create an array just to invoke the method but this too has a state inside of the item. Maybe it's better to rethink the delete use-case?

Reply:
If SaveState takes an extra ETag it keeps it simple.

Reply:
Not sure I follow. SaveState takes single parameter right now `State`

```go
State struct {
    StoreName string
    States    []*StateItem
}
```

If you add Etag to the signature, what would it do? Each one of the state items has its own Etag already, no?

Reply:
sorry, I meant SaveStateWithData

Reply:
No patience required, this looks way better now. These massive PRs aren't easy to land

Reply:
Added etag input param in the SaveStateWithData method